### PR TITLE
Make 'allAlpha' a special feature gate

### PR DESCRIPTION
**Pull request by thockin**
Rather than making all caller check both allAlpha and their own flag, make `allAlpha` set all of the alpha gates explicitly.

This is hard to test because of the globalness.  I will follow this commit with a new one to add some way to test, but I wanted to float this design

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31140)

### CODE REVIEW

**Comment by jlowdermilk** [comment]
I like having it default all flags first.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/util/config/feature_gate.go, line 53 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140#-KPnS9HCUPQniQarhPeI:-KPnS9HCUPQniQarhPeJ:bhg8rrh) ([raw file](https://github.com/kubernetes/kubernetes/blob/d4a8b1f9dec1e9bfca7668ffdb2aed5b244f3f59/pkg/util/config/feature_gate.go#L53)):*
> ```Go
> 	}
> 
> 	// Special handling for a few gates.
> ```

I don't expect we'll do special handling for any other flags. The above comment is enough for documentation, just do the defaulting inline in Set()

---

*[pkg/util/config/feature_gate.go, line 123 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140#-KPnSqa8a9nI8s0R0QLh:-KPnSqa8a9nI8s0R0QLi:bcropcl) ([raw file](https://github.com/kubernetes/kubernetes/blob/d4a8b1f9dec1e9bfca7668ffdb2aed5b244f3f59/pkg/util/config/feature_gate.go#L123)):*
> ```Go
> 		if fn, found := specialFeatures[k]; found {
> 			fn(f, boolValue)
> 		}
> ```

This should happen outside the loop, after all features are parsed. It should go through all keys in `knownFeatures`, and set the value iff it isn't already present in `f.features`. That way it sets default, but feature flag can override.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140#-:-KPnSTBJ9HisCH8mg1xU:bi98ba4)*

**Comment by thockin** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/util/config/feature_gate.go, line 53 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140#-KPnS9HCUPQniQarhPeI:-KPnbPK93Ryc020EXRc_:b4ba6j4) ([raw file](https://github.com/kubernetes/kubernetes/blob/d4a8b1f9dec1e9bfca7668ffdb2aed5b244f3f59/pkg/util/config/feature_gate.go#L53)):*
<details><summary><i>Previously, jlowdermilk (Jeff Lowdermilk) wrote…</i></summary>
> I don't expect we'll do special handling for any other flags. The above comment is enough for documentation, just do the defaulting inline in Set()
</details>
can I beg otherwise?  This is a little over-engineered today, but that's not so bad, should we ever want an allBeta flag or allNetwork flag or something.

---

*[pkg/util/config/feature_gate.go, line 123 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140#-KPnSqa8a9nI8s0R0QLh:-KPnboDSpmzFrHw60Bng:b-c5y63j) ([raw file](https://github.com/kubernetes/kubernetes/blob/d4a8b1f9dec1e9bfca7668ffdb2aed5b244f3f59/pkg/util/config/feature_gate.go#L123)):*
<details><summary><i>Previously, jlowdermilk (Jeff Lowdermilk) wrote…</i></summary>
> This should happen outside the loop, after all features are parsed. It should go through all keys in `knownFeatures`, and set the value iff it isn't already present in `f.features`. That way it sets default, but feature flag can override.
</details>
why is that better?  setting it here means it can be specified multiple times, which is weird but semantically consistent.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140)*

**Comment by thockin** [comment]
now with tests

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140#-:-KPndUt102Gpqpu2ORQu:b-twipka)*

**Comment by jlowdermilk** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/util/config/feature_gate.go, line 53 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140#-KPnS9HCUPQniQarhPeI:-KPncP5HQRK2oDsJd5Bw:b-daj5ev) ([raw file](https://github.com/kubernetes/kubernetes/blob/d4a8b1f9dec1e9bfca7668ffdb2aed5b244f3f59/pkg/util/config/feature_gate.go#L53)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary>
> can I beg otherwise?  This is a little over-engineered today, but that's not so bad, should we ever want an allBeta flag or allNetwork flag or something.
</details>
Sure, I don't feel strongly about it.

---

*[pkg/util/config/feature_gate.go, line 123 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140#-KPnSqa8a9nI8s0R0QLh:-KPncSHyI3lKjhPnJVdI:bu2liro) ([raw file](https://github.com/kubernetes/kubernetes/blob/d4a8b1f9dec1e9bfca7668ffdb2aed5b244f3f59/pkg/util/config/feature_gate.go#L123)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary>
> why is that better?  setting it here means it can be specified multiple times, which is weird but semantically consistent.
</details>
It needs to happen outside the loop otherwise the behavior changes depending on where in the string it's specified. I.e. `--feature-gates=FooBar=false,AllAlpha=true` != `--feature-gates=AllAlpha=true,FooBar=false`. You have to finish parsing all keys before doing any special behavior that changes `f.features`

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140)*

**Comment by thockin** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/util/config/feature_gate.go, line 123 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140#-KPnSqa8a9nI8s0R0QLh:-KPnd_xT8nQZq4UMfv3_:bynq4r4) ([raw file](https://github.com/kubernetes/kubernetes/blob/d4a8b1f9dec1e9bfca7668ffdb2aed5b244f3f59/pkg/util/config/feature_gate.go#L123)):*
<details><summary><i>Previously, jlowdermilk (Jeff Lowdermilk) wrote…</i></summary>
> It needs to happen outside the loop otherwise the behavior changes depending on where in the string it's specified. I.e. `--feature-gates=FooBar=false,AllAlpha=true` != `--feature-gates=AllAlpha=true,FooBar=false`. You have to finish parsing all keys before doing any special behavior that changes `f.features`
</details>
I was considering that to be expected behavior, I even added tests for it.  Too confusing?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140)*

**Comment by jlowdermilk** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/util/config/feature_gate.go, line 123 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140#-KPnSqa8a9nI8s0R0QLh:-KPnhiSpxLQ5oE6q56nX:bbyxiqb) ([raw file](https://github.com/kubernetes/kubernetes/blob/d4a8b1f9dec1e9bfca7668ffdb2aed5b244f3f59/pkg/util/config/feature_gate.go#L123)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary>
> I was considering that to be expected behavior, I even added tests for it.  Too confusing?
</details>
Way too confusing, imo. It's fine if specifying the same key=value twice picks the last one, but if you specify 2 different pairs in different orders, you should get the same behavior.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140)*

**Comment by thockin** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/util/config/feature_gate.go, line 123 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140#-KPnSqa8a9nI8s0R0QLh:-KPnjVS7PD2hI9JzP15Y:b7944at) ([raw file](https://github.com/kubernetes/kubernetes/blob/d4a8b1f9dec1e9bfca7668ffdb2aed5b244f3f59/pkg/util/config/feature_gate.go#L123)):*
<details><summary><i>Previously, jlowdermilk (Jeff Lowdermilk) wrote…</i></summary>
> Way too confusing, imo. It's fine if specifying the same key=value twice picks the last one, but if you specify 2 different pairs in different orders, you should get the same behavior.
</details>
will retool, stand by.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140)*

**Comment by thockin** [comment]
pushed with fixes more or less as requested.  Semantic is cleaner.  PTAL 

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140#-:-KPnn6p5BTdp-Qc99iq8:bt3d4z3)*

**Comment by girishkalele** [comment]
Reviewed 1 of 2 files at r3.
Review status: 1 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140)*

**Comment by jlowdermilk** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140#-:-KPnw0QnNWuRXqQMwxLO:bnfp4nl)*

**Comment by jlowdermilk** [comment]
Looks good, thanks. Feel free to add label after squashing.

**Comment by girishkalele** [comment]
Will rebase on top of this pending PR - there should be no changes to my application code - will have some conflicts with the new feature flag.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140#-:-KPnxQOO5KMQfm2rv8Qb:bllmhyb)*

**Comment by girishkalele** [comment]
Reviewed 1 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140)*

**Comment by thockin** [comment]
I squashed so it might make your rebase eventually conflict, but no code changes.

+lgtm 

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140#-:-KPo24zs3U3Jx227NKAp:bcwy6a9)*

File: pkg/util/config/feature_gate.go [hunk 1]
```diff
@@ -31,63 +31,93 @@ const (
...
+	beta  = prerelease("BETA")
+	ga    = prerelease("")
+)
+
 // FeatureGate parses and stores flag gates for known features from
 // a string like feature1=true,feature2=false,...
 type FeatureGate interface {
 	AddFlag(fs *pflag.FlagSet)
-	// owner: @jlowdermilk
-	// alpha: v1.4
```
Comment (jlowdermilk) [inline pkg/util/config/feature_gate.go]: Can you put this comment back in, maybe on the feature key const? Want to keep pattern of documenting owners

Comment (girishkalele) [inline pkg/util/config/feature_gate.go]: I noticed AllAlpha disappeared from the FeatureGate interface.
Do we have a use case for applications calling it directly ? Thinking out loud, I think there isn't a use case, applications should be querying their specific feature gate only.

Comment (thockin) [inline pkg/util/config/feature_gate.go]: I removed teh AllAlpha() method, but I left it documented below so the pattern is (hopefully) established

**Comment by thockin** [comment]
@k8s-bot test this: github issue #IGNORE

**Comment by thockin** [comment]
I added another commit to not panic if the flag was not parsed.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140#-:-KPop5bkMgp9oNeEr6Ww:b-9zvot0)*

**Comment by jlowdermilk** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r4.
Review status: 1 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140#-:-KPoqPAsdABvNkNXRAfJ:bnfp4nl)*

**Comment by thockin** [comment]
Girish did an over-the-shoulder code review. 

-lgtm 
+lgtm

---

Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31140#-:-KPoqHhWZ9XZsRFYtQjt:b-rzhj3)*

**Comment by thockin** [comment]
@k8s-bot test this: github issue #IGNORE

**Pull request merged**